pyensembl/transcript.py

```diff
@@ -69,46 +69,40 @@ def __len__(self):
         """
         Length of a transcript is the sum of its exon lengths
         """
-        # cache the length once you compute it
-        if not hasattr(self, "_length"):
-            self._length = sum(len(exon) for exon in self.exons)
-        return self._length
+        return len(self.sequence)
 
-    @property
+    @memoized_property
     def exons(self):
-        if not hasattr(self, "_exons"):
-            columns = ['exon_number', 'exon_id']
-            results = self.db.query(
-                columns,
-                filter_column='transcript_id',
-                filter_value=self.id,
-                feature='exon')
-
-             # fill this list in its correct order (by exon_number) by using
-             # the exon_number as a 1-based list offset
-            exons = [None] * len(results)
+        columns = ['exon_number', 'exon_id']
+        results = self.db.query(
+            columns,
+            filter_column='transcript_id',
+            filter_value=self.id,
+            feature='exon')
 
-            for entry in results:
-                exon_number, exon_id = entry
+         # fill this list in its correct order (by exon_number) by using
+         # the exon_number as a 1-based list offset
+        exons = [None] * len(results)
 
-                exon = Exon(exon_id, self.db)
-                exon_number = int(exon_number)
-                assert exon_number >= 1, "Invalid exon number: %s" % exon_number
-                assert exon_number <= len(exons), \
-                    "Invalid exon number: %s (max expected = %d)" % (
-                        exon_number, len(exons))
+        for entry in results:
+            exon_number, exon_id = entry
```

Comment:
Just FWIW you could do this destructuring in the loop, `for exon_number, exon_id in results:`